File: store/src/page_cache.rs
use std::{
    alloc, collections,
    ops::{Index, IndexMut},
    ptr, slice,
};

use format::{Format, page};

use crate::io;

pub struct PageCache {
    pub buf_pool: Vec<PageBuffer>,
    pub free_list: Vec<usize>,
    pub id_map: collections::BTreeMap<format::PageId, usize>,
    pub hits: Vec<[u64; 2]>,
    pub hit: u64,
    pub free_cap_target: usize,
}
impl PageCache {
    pub fn new(page_size: usize, buf_pool_size: usize, free_cap_target: usize) -> Self {
        Self {
            buf_pool: Vec::from_iter((0..buf_pool_size).map(|_| PageBuffer::new(page_size))),
            free_list: Vec::from_iter(0..buf_pool_size),
            id_map: collections::BTreeMap::new(),
            hits: vec![[u64::MAX, 0]; buf_pool_size],
            hit: 0,
            free_cap_target,
        }
    }

    /// returns the index in the buffer pool for `pid`, if it exists.
    /// also records a "hit" for eviction priority tracking purposes.
    pub fn get(&mut self, pid: format::PageId) -> Option<usize> {
        match self.id_map.get(&pid) {
            Some(&idx) => {
                self.hits[idx][1] = self.hits[idx][0];
                self.hits[idx][0] = self.hit;
                self.hit += 1;
                Some(idx)
            }
            None => None,
        }
    }

    /// insert `pid` at `idx`.
    /// actual modification of the [`PageBuffer`] at `idx` is done separately.
    /// also records a "hit" for eviction priority tracking purposes.
    pub fn insert(&mut self, pid: format::PageId, idx: usize) {
        if let Some(_) = self.id_map.insert(pid, idx) {
            panic!()
        }
        self.hits[idx][0] = self.hit;
        self.hits[idx][1] = 0;
        self.hit += 1;
    }

    /// remove `pid` from cache.
    pub fn remove(&mut self, pid: format::PageId) {
        self.id_map.remove(&pid).unwrap();
    }

    /// pops a free slot in the buffer pool from the free list, panics if free list is empty
    pub fn pop_free(&mut self) -> usize {
        self.free_list.pop().unwrap()
    }

    /// run eviction if necessary
    pub fn tick<IO: io::IOFace>(&mut self, page_store: &mut io::PageStore, io: &mut IO) {
        let mut evict_cands = itertools::Itertools::sorted_by(
            self.id_map
                .iter()
                .map(|(pid, idx)| (self.hits[*idx][0] - self.hits[*idx][1], (*pid, *idx))),
            |a, b| a.0.cmp(&b.0),
        )
        .rev()
        .map(|(_, out)| out);
        while self.free_list.len() < self.free_cap_target {
            let (pid, idx) = evict_cands.next().unwrap();
            // for now we'll just evict the best candidates, but we may have some constraints
            // later from things like txns that prevent us from evicting certain pages
            let page = &mut self.buf_pool[idx];
            let total = page.cap - page.top;
            let chunk = page.flush();
            let base = chunk.len() == total;

            page_store.write_chunk(pid, chunk, io, base);

            // cleanup
            page.clear();
            self.free_list.push(idx);
            self.remove(pid);
        }
    }
}
impl Index<usize> for PageCache {
    type Output = PageBuffer;
    fn index(&self, index: usize) -> &Self::Output {
        self.buf_pool.index(index)
    }
}
impl IndexMut<usize> for PageCache {
    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
        self.buf_pool.index_mut(index)
    }
}

pub struct PageBuffer {
    ptr: ptr::NonNull<u8>,
    pub top: usize,
    pub cap: usize,
    pub flush: usize,
}
impl PageBuffer {
    pub fn new(cap: usize) -> Self {
        let layout = alloc::Layout::array::<u8>(cap).unwrap();
        let ptr = match ptr::NonNull::new(unsafe { alloc::alloc_zeroed(layout) }) {
            Some(p) => p,
            None => alloc::handle_alloc_error(layout),
        };
        Self {
            ptr,
            cap,
            top: cap,
            flush: cap,
        }
    }

    pub fn read(&self) -> page::Page<'_> {
        page::Page::from(unsafe {
            slice::from_raw_parts(self.ptr.add(self.top).as_ptr(), self.cap - self.top)
        })
    }
    pub fn write(&mut self, op: &page::PageOp) -> bool {
        let len = op.len();

        if len > self.top {
            return false;
        }

        op.write_to_buf(unsafe {
            slice::from_raw_parts_mut(self.ptr.add(self.top - len).as_ptr(), len)
        });
        self.top -= len;

        true
    }

    #[inline]
    pub fn raw_buffer_mut(&mut self) -> &mut [u8] {
        unsafe { slice::from_raw_parts_mut(self.ptr.as_ptr(), self.cap) }
    }
    pub fn clear(&mut self) {
        self.raw_buffer_mut().fill(0);
        self.top = self.cap;
        self.flush = self.cap;
    }
    pub fn flush(&mut self) -> &[u8] {
        let out =
            &(unsafe { slice::from_raw_parts(self.ptr.as_ptr(), self.cap) })[self.top..self.flush];
        self.flush = self.top;
        out
    }
}
impl Drop for PageBuffer {
    fn drop(&mut self) {
        unsafe {
            alloc::dealloc(
                self.ptr.as_ptr(),
                alloc::Layout::array::<u8>(self.cap).unwrap(),
            )
        }
    }
}

File: src/PageCache.zig
//! ## TODO
//! - eviction (when we figure out io)

const std = @import("std");
const mem = std.mem;
const testing = std.testing;

const format = @import("format.zig");

buf_pool: []Buffer,
free_list: std.ArrayListUnmanaged(usize),
id_map: std.AutoHashMapUnmanaged(u64, usize),

hits: [][2]u64,
hit: u64,
free_cap_target: usize,

const Self = @This();

pub fn init(
    page_size: usize,
    pool_size: usize,
    free_cap_target: usize,
    allocator: mem.Allocator,
) Self {
    const buf = allocator.alloc(u8, page_size * pool_size) catch unreachable;
    var buf_pool = allocator.alloc(Buffer, pool_size) catch unreachable;
    var free_list = std.ArrayListUnmanaged(usize).initCapacity(
        allocator,
        pool_size,
    ) catch unreachable;
    var id_map = std.AutoHashMapUnmanaged(u64, usize).empty;
    id_map.ensureTotalCapacity(
        allocator,
        @intCast(pool_size),
    ) catch unreachable;

    var hits = allocator.alloc([2]u64, pool_size) catch unreachable;

    for (0..pool_size) |i| {
        buf_pool[i] = Buffer.init(buf[i * page_size .. (i + 1) * page_size]);
        free_list.appendAssumeCapacity(i);
        hits[i] = .{ std.math.maxInt(u64), 0 };
    }

    return Self{
        .buf_pool = buf_pool,
        .free_list = free_list,
        .id_map = id_map,

        .hits = hits,
        .hit = 1,
        .free_cap_target = free_cap_target,
    };
}

pub fn deinit(self: *Self, allocator: mem.Allocator) void {
    var og_buf = self.buf_pool[0].buf;
    og_buf.len = og_buf.len * self.buf_pool.len;
    allocator.free(og_buf);
    allocator.free(self.buf_pool);

    self.free_list.deinit(allocator);
    self.id_map.deinit(allocator);

    allocator.free(self.hits);
}

pub fn get(self: *Self, pid: u64) ?usize {
    if (self.id_map.get(pid)) |idx| {
        self.hits[idx][1] = self.hits[idx][0];
        self.hits[idx][0] = self.hit;
        self.hit += 1;
        return idx;
    } else {
        return null;
    }
}

pub fn insert(self: *Self, pid: u64, idx: usize) void {
    if (self.id_map.fetchPutAssumeCapacity(pid, idx)) {
        unreachable;
    }
    self.hits[idx][0] = self.hit;
    self.hits[idx][1] = 0;
    self.hit += 1;
}

pub fn remove(self: *Self, pid: u64) void {
    if (!(self.id_map.fetchRemove(pid))) {
        unreachable;
    }
}

pub fn pop_free(self: *Self) usize {
    if (self.free_list.pop()) |idx| {
        return idx;
    } else {
        unreachable;
    }
}

pub const Buffer = struct {
    buf: []u8,
    top: usize,

    const _Self = @This();

    const Error = error{
        BufFull,
    };

    pub fn init(buf: []u8) _Self {
        return _Self{ .buf = buf, .top = buf.len };
    }

    pub fn reset(self: *_Self) void {
        @memset(self.buf, 0);
        self.top = self.buf.len;
    }

    pub fn write(
        self: *_Self,
        comptime W: type,
        w: *const W,
    ) _Self.Error!void {
        const size = w.size();
        if (size > self.top) {
            return _Self.Error.BufFull;
        }
        w.serialize(self.buf[self.top - size .. self.top]);
        self.top -= size;
    }

    pub fn read(self: *const _Self) format.Page {
        return format.Page.fromBytes(self.buf[self.top..]);
    }
};

test {
    var cache = Self.init(1024, 1024, 64, testing.allocator);
    defer cache.deinit(testing.allocator);
}
